### PR TITLE
build with openCV4

### PR DESCRIPTION
**Pull request by doisyg**
I needed image_view on foxy/ubuntu focal, so I needed image_pipeline to build with OpenCV4. 
As I understand it the version of openCV for foxy is 4, with no back support of earlier versions. See for instance the  vision_opencv packages https://github.com/ros-perception/vision_opencv/tree/ros2

I have only tested image_view which works fine.

### CODE REVIEW

**Comment by doisyg** [comment]
CI is failing but I guess it is because it is testing for dashing.
Shouldn't the ros2 branch target the next release (ie foxy) as in most package ?
"ros2: points to the next unreleased ROS 2 turtle, currently Foxy. "

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
@doisyg can you update CI to run Foxy / ROS2 master?

**Comment by doisyg** [comment]
Sure but I am not familiar with CircleCI, should changing `eloquent` for `foxy` in image_pipeline/.circleci/config.yml be enough? I seems that it uses the *turtle*-ros-base docker, but is there a foxy-ros-base out there already ?

**Comment by SteveMacenski** [comment]
Well, right now its Dashing :laughing: https://github.com/ros-perception/image_pipeline/blob/ros2/.circleci/config.yml#L5 but more or less. Right now your PR fails because openCV4 isn't in dashing. There's also a ROS2 nightly dockerfile that could be used as well instead.

**Comment by doisyg** [comment]
Indeed i just found out about osrf/ros2:nightly, but `image_pipeline` depends on `image_common` for `image_transport` and `image_geometry` and it is not released yet ... though there must be a way

**Comment by SteveMacenski** [comment]
Typically how that could be dealt with is via building on nightly, but in the circle CI file, have it build this repo with those others repos using a `dependencies.repos` file (e.x. https://github.com/ros-planning/navigation2/blob/master/tools/ros2_dependencies.repos) for temporary use. Or we could just wait until a docker image is available.

**Comment by JWhitleyWork** [comment]
@doisyg / @SteveMacenski I've created some Foxy Docker images which you can use. Just use the image `jwhitleywork/ros:foxy-ros-base-focal`. However, this isn't just a CI issue - There were significant changes (especially in linting) between Dashing, Eloquent, and Foxy. Because of this, we will need to maintain different branches for these three.

@SteveMacenski I feel like we've discussed this before on this repo. Thoughts on targeting `foxy` with the `ros2` branch while branching `dashing` and `eloquent` at the current state?

**Comment by SteveMacenski** [comment]
Yes, ros2 is targeting ros2 master, which is currently foxy. Before this PR goes in, we should branch the current state of ros2 as `dashing -devel` to cover dashing and eloquent (I don't think anything should be an issue with those as long as there's no TF... which I don't think there is).

**Comment by doisyg** [comment]
Last commits for CI.  Use `jwhitleywork/ros:foxy-ros-base-focal`  and naively pull the `image_common` and `vision_opencv` packages in the same ws. Strangely I had to install with apt `libgtk2.0-dev` which was not picked up by rosdep.
Compilation is okay, it will need more work for the tests (and I have no time nor the proper expertise for it)

**Comment by SteveMacenski** [comment]
@doisyg I think that's because GTK3.0 is what's used now. Can you run `ament_uncrustify --reformat <path>` over the repo? That will fix the linter errors (or at least many of them).

@JWhitleyWork Should any of the patches from Noetic be made here for python3? I'm seeing some test failures.

**Comment by JWhitleyWork** [comment]
@SteveMacenski The GTK 3 change was only made on the Noetic branch. It should be applied here too.

**Comment by doisyg** [comment]
Switched to gtk3 in CMakeLists.txt and package.xml as in noetic and removed the almost empty ifdef. It seems gtk is not used at all in the code, should we remove it completely ? Or should we left it if it is needed to implement this in the future in Foxy:
https://github.com/ros-perception/image_pipeline/blob/f13c10f643501ecfa612fecaf2944eb181f36888/image_view/src/nodes/stereo_view.cpp#L52-L62

Or maybe this workaround is not needed anymore on Foxy platforms

Also ran `ament_uncrustify --reformat`

**Comment by SteveMacenski** [comment]
There are still a lot of test failures in CI - It looks like to me * that all of those errors are from image_common or the other repos we pulled in to do CI. I think that can be resolved by adding `--packages-select camera_calibration depth_image_proc image_proc image_pipeline image_publisher image_rotate image_view stereo_image_proc` to the colcon test line. That way it skips running tests over our dependencies which appear to not be linted (and broken)

This would be a nice place for special metapackage depth rules; test all in metapackage recursively for depth 1

**Comment by doisyg** [comment]
> There are still a lot of test failures in CI - It looks like to me * that all of those errors are from image_common or the other repos we pulled in to do CI. I think that can be resolved by adding `--packages-select camera_calibration depth_image_proc image_proc image_pipeline image_publisher image_rotate image_view stereo_image_proc` to the colcon test line. That way it skips running tests over our dependencies which appear to not be linted (and broken)
> 

Indeed, looks like it passes the CI now

**Comment by JWhitleyWork** [comment]
@SteveMacenski `dashing` (dropped `-devel` consistent with newer branches on this repo) has been created from the current state of `ros2`. Somehow, we never released `image_pipeline` for Dashing - do you want me to go ahead and do that?

Regardless, this MR can be merged.

**Comment by SteveMacenski** [comment]
Sure, lets release it

**Comment by doisyg** [comment]
Would be nice to release it for Foxy too

**Comment by SteveMacenski** [comment]
@doisyg definitely for foxy, its only been in the last month or so we've had a full port. There's still cleanup to do, but I think its in a state that can be called "good enough for now". Still some functionality like reconfiguring parameters and only publishing when someone is subscribed that aren't implemented right now. 

We still need at minimum vision_opencv and image_common to be released before we can in Foxy

**Comment by ivanpauno** [comment]
@SteveMacenski Is `image_pipeline` supporting old ros2 distros (Dashing and Eloquent)?

Ubuntu 18.04 only have debians for OpenCV 3, and installing OpenCV 4 on it is an extra effort.
I don't see a branch compatible with OpenCV 3.

**Comment by ivanpauno** [comment]
> I don't see a branch compatible with OpenCV 3.

I'm sorry, I see there's a `Dashing` branch. Is that one compatible with `Eloquent`, or should one branch for it be created?

**Comment by SteveMacenski** [comment]
Dashing branch serves both dashing and eloquent

**Comment by ivanpauno** [comment]
> Dashing branch serves both dashing and eloquent

Thanks!

**Comment by JWhitleyWork** [comment]
> Sure, lets release it

Would you mind pushing the release? I'm busy as hell right now.

**Comment by SteveMacenski** [comment]
I'm pretty sure we already did.......


.... or was that Noetic? jeeze 2 different releases at once is really not a good habit to get into. OK, I'll figure it out and add to my mountain of releases I was planning on doing on Friday.

**Comment by SteveMacenski** [comment]
https://github.com/ros/rosdistro/pull/25137

**Comment by SteveMacenski** [comment]
Memory serves that failures due to
```
dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
E: Building failed
Traceback (most recent call last):
  File "/tmp/ros_buildfarm/ros_buildfarm/binarydeb_job.py", line 138, in build_binarydeb
    subprocess.check_call(cmd, cwd=source_dir)
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['apt-src', 'build', 'ros-foxy-image-view']' returned non-zero exit status 1.
# END SUBSECTION
```

are internal build farm things, correct? Just making sure I shouldn't be doing anything

**Comment by JWhitleyWork** [comment]
@SteveMacenski Can you link to the build where you're seeing this?

**Comment by SteveMacenski** [comment]
You mean you're not getting emails about it? Maybe we should add you to the package.xmls :wink: http://build.ros2.org/job/Fbin_uF64__image_view__ubuntu_focal_amd64__binary/2/console

**Comment by JWhitleyWork** [comment]
@SteveMacenski Check these out from the build log:

```
23:01:43 CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
23:01:43   Could NOT find Boost (missing: Boost_INCLUDE_DIR)
23:01:43 Call Stack (most recent call first):
23:01:43   /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
23:01:43   /usr/share/cmake-3.16/Modules/FindBoost.cmake:2179 (find_package_handle_standard_args)
23:01:43   CMakeLists.txt:22 (find_package)
```

`grep`-ing through the code (in a one-horse open slay), it looks like there are still a few references to Boost in some of the files but none of them have a Boost dependency in `package.xml`.

**Comment by SteveMacenski** [comment]
how did that work for the last release....

.... that's right, we haven't done a release yet for ROS2. This is going to be a really long process that I can't take on this week. It'll have to wait for next week. I think image_view is the only one failing though, or at least the only one I'm getting emails about.

**Comment by SteveMacenski** [comment]
I just checked, its only image view failing of the builds that have processed (some bin builds havent triggered yet). So it might not be as bad as I thought. I'll handle next week.

File: image_publisher/CMakeLists.txt [hunk 1]
```diff
@@ -29,15 +29,9 @@ set( DEPENDENCIES
...
-set(opencv_2_components core highgui)
-set(opencv_3_components core imgcodecs videoio)
 find_package(OpenCV REQUIRED COMPONENTS core)
 message(STATUS "opencv version ${OpenCV_VERSION}")
-if(OpenCV_VERSION VERSION_LESS "3.0.0")
-  find_package(OpenCV 2 REQUIRED COMPONENTS ${opencv_2_components})
-else()
-  find_package(OpenCV 3 REQUIRED COMPONENTS ${opencv_3_components})
-endif()
+find_package(OpenCV 4 REQUIRED COMPONENTS ${opencv_4_components})
```
Comment (dirk-thomas) [inline image_publisher/CMakeLists.txt:34]: The variable `opencv_4_components` doesn't seem to be set anywhere?

**Pull request merged**